metriq_gym/benchmarks/chsh.py

```diff
@@ -0,0 +1,266 @@
+
+The CHSH benchmark evaluates a quantum device's ability to produce entangled states and measure correlations that
+violate the CHSH inequality. The violation of this inequality indicates successful entanglement between qubits.
+"""
+
+from dataclasses import dataclass
+import networkx as nx
+import numpy as np
+from qbraid import QuantumDevice, QuantumJob, ResultData
+from qbraid.runtime import (
+    QiskitBackend,
+    BraketDevice,
+    IonQDevice,
+)
+import rustworkx as rx
+
+from qiskit import QuantumCircuit
+from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager
+from qiskit.result import marginal_counts, sampled_expectation_value
+
+from metriq_gym.benchmarks.benchmark import Benchmark, BenchmarkData
+
+
+class GraphColoring:
+    """A simple class containing graph coloring data.
+
+    Attributes:
+        num_nodes: Number of qubits (nodes) in the graph.
+        edge_color_map: Maps each edge index to a color (integer).
+        edge_index_map: Maps edge indices to actual qubit pairs.
+        num_colors: Total number of colors assigned in the graph.
+    """
+
+    def __init__(self, num_nodes: int, edge_color_map: dict, edge_index_map: dict) -> None:
+        self.edge_color_map = edge_color_map
+        self.edge_index_map = edge_index_map
+        self.num_colors = max(edge_color_map.values()) + 1
+        self.num_nodes = num_nodes
+
+
+@dataclass
+class CHSHData(BenchmarkData):
+    """Data class to store CHSH benchmark metadata.
+
+    Attributes:
+        shots: Number of shots per quantum circuit execution.
+        num_qubits: Number of qubits in the quantum device.
+        topology_graph: Graph representing the device topology (optional).
+        coloring: Coloring information for circuit partitioning (optional).
+    """
+
+    shots: int
+    num_qubits: int
+    topology_graph: nx.Graph | None = None
+    coloring: GraphColoring | None = None
+
+
+def device_coloring(device: QuantumDevice) -> GraphColoring:
+    """Performs graph coloring for a quantum device's topology.
+
+    The goal is to assign colors to edges such that no two adjacent edges have the same color.  This ensures independent
+    CHSH experiments can be run in parallel. Identifies qubit pairs (edges) that can be executed without interference.
+    These pairs are grouped by "color." The coloring reduces the complexity of the benchmarking process by organizing
+    the graph into independent sets of qubit pairs.
+
+    Args:
+        device: The quantum device.
+
+    Returns:
+        GraphColoring: An object containing the coloring information.
+    """
+    match device:
+        case IonQDevice():
+            # IonQ devices use a fully connected (complete) graph.
+            topology_graph = nx.complete_graph(device.num_qubits)
+            edge_color_map = {i: 0 for i in range(topology_graph.number_of_edges())}
+            topology_graph = rx.networkx_converter(topology_graph)
```

Comment:
It doesn't look like there is anything specific to IonQDevice here, we can have this case as a fallback for any device that doesn't expose a coupling map.